backends/dpdk/dpdkArch.cpp

```diff
@@ -3035,4 +3035,153 @@ std::vector<std::pair<cstring, const IR::Type *>>
+        auto typeArgs = new IR::Vector<IR::Type>();
+        typeArgs->push_back(new IR::Type_Name(IR::ID(newHeaderName)));
+        auto regRdArgs = new IR::Vector<IR::Argument>();
+
+        // Register index is always 0 as the ipsec port registers are all single location registers
+        regRdArgs->push_back(
+            new IR::Argument(new IR::Constant(IR::Type::Bits::get(32), IPSEC_PORT_REG_INDEX)));
+        /*  Set the ipsec output port based on packet direction and emit ipsec_hdr
+         *  if (ipsec_hdr.isValid())
+                 port_out = (direction == NET_TO_HOST) ?
+                            ipsec_inbound_port_out: ipsec_outbound_port_out;
+                 emit(ipsec_hdr);
+        */
+        auto regRdOutInbound = new IR::MethodCallExpression(
+            new IR::Member(new IR::PathExpression(IR::ID("ipsec_port_out_inbound")),
+                           IR::ID("read")),
+            regRdArgs);
+        auto regRdOutOutbound = new IR::MethodCallExpression(
+            new IR::Member(new IR::PathExpression(IR::ID("ipsec_port_out_outbound")),
+                           IR::ID("read")),
+            regRdArgs);
+
+        auto portOutInBound = new IR::AssignmentStatement(
+            c->body->srcInfo,
+            new IR::Member(new IR::PathExpression(IR::ID("m")),
+                           IR::ID("pna_main_output_metadata_output_port")),
+            regRdOutInbound);
+        auto portOutOutBound = new IR::AssignmentStatement(
+            c->body->srcInfo,
+            new IR::Member(new IR::PathExpression(IR::ID("m")),
+                           IR::ID("pna_main_output_metadata_output_port")),
+            regRdOutOutbound);
+        auto isNetToHost = new IR::Equ(c->body->srcInfo,
+                                       new IR::Member(new IR::PathExpression(IR::ID("m")),
+                                                      IR::ID("pna_main_input_metadata_direction")),
+                                       new IR::Constant(NET_TO_HOST));
+        auto isValid =
+            new IR::Member(c->body->srcInfo,
+                           new IR::Member(new IR::PathExpression(IR::ID("h")), IR::ID("ipsec_hdr")),
+                           IR::ID(IR::Type_Header::isValid));
+
+        auto ipsecIsValid =
+            new IR::MethodCallExpression(c->body->srcInfo, IR::Type::Boolean::get(), isValid);
+        auto ifTrue =
+            new IR::IfStatement(c->body->srcInfo, isNetToHost, portOutInBound, portOutOutBound);
+        IR::Statement *fillIpsecPortOut = new IR::IfStatement(ipsecIsValid, ifTrue, nullptr);
+        IR::Statement *ipsecEmit = new IR::MethodCallStatement(new IR::MethodCallExpression(
+            new IR::Member(new IR::PathExpression(pkt_out->name), IR::ID("emit")), typeArgs, args));
+        stmt.push_back(fillIpsecPortOut);
+        stmt.push_back(ipsecEmit);
+        return stmt;
+    };
+
+    for (auto kv : structure->deparsers) {
```

Comment:
this could be also done in a visitor

Reply:
Here, we need to insert this set of statements (from lambda function) in deparser. We have the information about which control blocks is/are deparser(s) stored in the "structure". This information is collected in the initial passes while inspecting the program as per specified architecture.  Hence, I am checking that while visiting P4Control. I am not able to think of any other way to achieve this here. Please suggest if I am missing something.